Pybind11 starting point

lamiller0 opened the pull request:
Address some of the feedback in #351 

Add a PYBIND11 option.

I attempted to put this into src/pybind11 instead of under src/python but I believe that the configuration under src/python/config was useful, and I didn't want to spend cycles figuring out what should be recreated.

lgritz commented on src/python/PyBindImath/PyImathVec.cpp:
```diff
@@ -0,0 +1,12 @@
+#include "PyImath.h"
```
Let's be sure to put the

```
//
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the OpenEXR Project.
//
```

at the head of all new source files.

lgritz commented on src/python/config/ModuleDefine.cmake:
```diff
@@ -100,8 +103,14 @@ function(PYIMATH_DEFINE_MODULE modname)
   list(APPEND libarglist PRIVATE_DEPS ${PYIMATH_CURMOD_PRIVATE_DEPS})
+
+  set(optional_boost "")
+  if (TARGET Boost::${PYIMATH_BOOST_PY_COMPONENT})
+    set(optional_boost Boost::${PYIMATH_BOOST_PY_COMPONENT})
+  endif()
```
I think the modern cmake approach is to use a generator expression like

    $<TARGET_NAME_IF_EXISTS:Boost::${PYIMATH_BOOST_PY_COMPONENT}>

and then you don't need the extra optional_boost variable or this `if` statement.

cary-ilm commented:
Does `src/python/PyImathTestC` build for you with `PYBIND11=ON`? I'm getting:
```
CMake Error at src/python/PyImathTest/CMakeLists.txt:23 (target_link_libraries):
  Target "PyImathTestC" links to:

    Imath::PyImath_Python3_10

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
```

If it's set up to configure with one of boost or pybind11 but not both, then PyImathTestC needs special attention. We'll also need a separate pybind11 test to start building up in parallel with the original, both python and C++.

I thought it would be better to have:

```
src
├── Imath
├── ImathTest
├── pybind11
│   ├── config
│   ├── PyImath
│   ├── PyImathTest
│   └── PyImathTestC
└── python
    ├── config
    ├── PyImath
    └── PyImathTest
    └── PyImathTestC
```

lgritz commented on src/python/config/ModuleDefine.cmake:
```diff
@@ -110,7 +119,7 @@ function(PYIMATH_DEFINE_MODULE modname)
         ${extraDeps}
         Python2::Python
-        Boost::${PYIMATH_BOOST_PY_COMPONENT}
+        ${optional_boost}
```
```suggestion
        $<TARGET_NAME_IF_EXISTS:Boost::${PYIMATH_BOOST_PY_COMPONENT}>
```

lgritz replied:
and in the other spots

lgritz reviewed (changes requested):
I made some very minor comments, but aside from that the general approach LGTM.

lamiller0 commented:
I've addressed most of the feedback.

For now with pybind11 I'm not including PyImathTest.  The config stuff is pretty dense, so I thought it would be better to just use it rather than copy it over to a different directory.

I'd also like to introduce python unit tests at some point soon, to help show the bindings are working in a similar way (where it makes sense).

cary-ilm commented:
We'll need the test right away, I wouldn't want anyone contributing additions without test coverage, and I'm curious exactly which functions the v3f implementation covers.  And regarding the config, my thought was to start a fresh config with a clean slate, without any reference to boost. But I'm happy to work on that later.

lamiller0 commented:
The pybind11 V3f as it is right now will be VERY disappointing.  😃 

I'll add in a python unit test framework like I have with Alembic which will run in either boost or pybind11 and try to craft the tests such that they will pass in both.

cary-ilm reviewed (approved):
Looks good to me as a starting point, thanks!

michdolan commented on src/python/PyBindImath/PyImathVec.cpp:
```diff
@@ -0,0 +1,17 @@
+void register_imath_vec(pybind11::module& m)
+{
+    // TODO fill in
+    pybind11::class_<Imath::V3f>(m, "V3f");
```
It would be great to implement the Python bindings so that there is a class hierarchy matching C++. So have a `Vec3` class, and the type-specific variations inheriting from that. This would improve type hinting and checking to support things like `isinstance(vector, Vec3)` and `def my_function(vector: Vec3) -> Matrix44:`. As an example of this approach, OCIO has implemented `Transform` and its subclasses in a similar way: 
https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/main/src/bindings/python/PyTransform.cpp

michdolan replied:
I'm happy to collaborate on this work if help is needed. I was getting ready to start experimenting myself with a pybind11 implementation of the bindings when I noticed this PR. Awesome to see this work is already started :)

cary-ilm replied:
Thanks, @michdolan, we're happy to have the help! And the class hierarchy sounds like a nice structure.

@lgritz, you've requested changes, are you OK with this as is for now?

I still think a separate `pybind11` subdirectory in parallel with `python` would make sense, since that will make room for the pybind11 tests, and also a fresh,  boost-free, pybind-specific config, but it's fine to deal with that later.

lamiller0 replied:
Yeah, I think separating it out in a parallel directory should be done later.  I'm currently planning on creating python tests that should pass with both the boost bindings and the pybind11 ones, and it will be more convenient for the time being to be under python.

lgritz reviewed (approved):
LGTM

cary-ilm commented:
Off to the races, thanks!
